File: src/IsbInterop/IsbInteropConfiguration.vb
Imports System.ComponentModel
Imports System.Configuration
Imports System.IO

Public NotInheritable Class IsbInteropConfiguration

  #Region "Поля и свойства"

  ''' <summary>
  ''' Таймаут выполнения метода IS-Builder, секунд.
  ''' </summary>
  Public Shared ReadOnly Property IsbMethodExecutionTimeout As Integer
    Get
      Return _isbMethodExecutionTimeout.Value
    End Get
  End Property

  Private Shared ReadOnly _isbMethodExecutionTimeout As New Lazy(Of Integer)(
    Function()
      Const defaultMethodExecutionTimeout = 60
      Return GetAppSettingOrDefault("IsbMethodExecutionTimeout", defaultMethodExecutionTimeout, Function(t) t > 0)
    End Function, True)

  ''' <summary>
  ''' Таймаут создания приложения IS-Builder, секунд.
  ''' </summary>
  Public Shared ReadOnly Property IsbAppCreationTimeout As Integer
    Get
      Return _isbAppCreationTimeout.Value
    End Get
  End Property

  Private Shared ReadOnly _isbAppCreationTimeout As New Lazy(Of Integer)(
    Function()
      Const defaultIsbAppCreationTimeout = 300
      Return GetAppSettingOrDefault("IsbAppCreationTimeout", defaultIsbAppCreationTimeout, Function(t) t > 0)
    End Function, True)

  ''' <summary>
  ''' Конфигурация сборки IsbInterop.
  ''' </summary>
  Private Shared ReadOnly StandaloneConfiguration As New Lazy(Of Configuration)(AddressOf TryGetStandaloneConfiguration, True)

  #End Region

  #Region "Методы"

    ''' <summary>
  ''' Шифрует строки подключения.
  ''' </summary>
  Public Shared Sub EncryptConnectionStrings()
    Dim configuration = StandaloneConfiguration.Value

    If configuration Is Nothing Then
      Try
        configuration = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None)
      Catch generatedExceptionName As ConfigurationErrorsException
        Return
      End Try
    End If

    Dim configSection = configuration.ConnectionStrings

    If configSection Is Nothing Then
      Return
    End If

    Try
      If Not configSection.IsReadOnly() AndAlso Not configSection.ElementInformation.IsLocked AndAlso Not configSection.SectionInformation.IsLocked Then
        If Not configSection.SectionInformation.IsProtected Then
          configSection.SectionInformation.ProtectSection("DataProtectionConfigurationProvider")
        End If

        configSection.SectionInformation.ForceSave = True

        configuration.Save()
      End If
      ' ReSharper disable once EmptyGeneralCatchClause
    Catch generatedExceptionName As Exception
    End Try
  End Sub

    ''' <summary>
  ''' Получает настройку секции appSettings.
  ''' </summary>
  ''' <typeparam name="T">Тип результата.</typeparam>
  ''' <param name="settingName">Имя настройки/</param>
  ''' <param name="result">Результат.</param>
  ''' <returns>True, если удалось получить настройку, иначе false.</returns>
  Public Shared Function TryGetAppSetting(Of T)(settingName As String, ByRef result As T) As Boolean
    If settingName Is Nothing Then
      Throw New ArgumentNullException("settingName")
    End If

    Dim settingValue As String = Nothing
    Dim configuration = StandaloneConfiguration.Value

    If configuration IsNot Nothing Then
      Dim setting = configuration.AppSettings.Settings(settingName)
      If setting IsNot Nothing Then
        settingValue = setting.Value
      End If
    Else
      settingValue = ConfigurationManager.AppSettings(settingName)
    End If

    If settingValue IsNot Nothing Then
      Return TryConvert(settingValue, result)
    Else
      Return False
    End If
  End Function

  ''' <summary>
  ''' Получает строку подключения.
  ''' </summary>
  ''' <returns>Настройки строки подключения.</returns>
  Friend  Shared Function GetConnectionString() As String
    Dim connectionStringSettings As ConnectionStringSettings
    Dim configuration = StandaloneConfiguration.Value

    If configuration IsNot Nothing Then
      connectionStringSettings = configuration.ConnectionStrings.ConnectionStrings("IsbInterop")
    Else
      connectionStringSettings = ConfigurationManager.ConnectionStrings("IsbInterop")
    End If

    Dim connectionString = If(connectionStringSettings IsNot Nothing, connectionStringSettings.ConnectionString, Nothing)

    Return connectionString
  End Function

  ''' <summary>
  ''' Получает конфигурацию сборки IsbInterop.
  ''' </summary>
  ''' <returns>Конфигурация, если она существует, иначе null.</returns>
  Private Shared Function TryGetStandaloneConfiguration() As Configuration
    Dim currentAssembly = GetType(IsbApplicationManager).Assembly
    Dim assemblyPath = New Uri(currentAssembly.CodeBase).LocalPath
    Dim assemblyPathWithoutExtenstion = assemblyPath.Substring(0, assemblyPath.Length - 4)
    Dim configPath = String.Format("{0}.dll.config", assemblyPathWithoutExtenstion)

    ' Проверяем, что файл существует, т.к.метод OpenMappedExeConfiguration всегда создает объект конфигурации.
    If Not File.Exists(configPath) Then
      Return Nothing
    End If

    Dim configMap = New ExeConfigurationFileMap()
    configMap.ExeConfigFilename = configPath

    Dim configuration As Configuration = Nothing

    Try
      configuration = ConfigurationManager.OpenMappedExeConfiguration(configMap, ConfigurationUserLevel.None)
      ' We could log.trace this exception if we have logger.
    Catch generatedExceptionName As ConfigurationErrorsException
    End Try

    Return configuration
  End Function

  ''' <summary>
  ''' Конвертирует строку в заданный тип.
  ''' </summary>
  ''' <typeparam name="T">Тип результата.</typeparam>
  ''' <param name="source">Исходное значение.</param>
  ''' <param name="result">Результат.</param>
  ''' <returns>Экземпляр типа.</returns>
  Private Shared Function TryConvert(Of T)(source As String, ByRef result As T) As Boolean
    Dim typeConverter = TypeDescriptor.GetConverter(GetType(T))

    If typeConverter.CanConvertFrom(GetType(String)) Then
      Try
        result = DirectCast(typeConverter.ConvertFrom(source), T)
        Return True
        ' ReSharper disable once EmptyGeneralCatchClause
      Catch ex As Exception
      End Try
    End If

    Return False
  End Function

  ''' <summary>
  ''' Получает значение настройки из секции AppSettings.
  ''' </summary>
  ''' <typeparam name="T">Тип значения.</typeparam>
  ''' <param name="setting">Настройка.</param>
  ''' <param name="defaultValue">Значение по умолчанию.</param>
  ''' <param name="predicate">Предикат.</param>
  ''' <returns>Значение настройки.</returns>
  Private Shared Function GetAppSettingOrDefault(Of T)(setting As String, defaultValue As T,
                                                       Optional predicate As Func(Of T, Boolean) = Nothing) As T
    Dim value As T = defaultValue

    If TryGetAppSetting(setting, value) Then
      If predicate IsNot Nothing AndAlso predicate(value) Then
        Return value
      End If
    End If

    Return defaultValue
  End Function

  #End Region

  #Region "Конструкторы"

  ''' <summary>
  ''' Приватный конструктор.
  ''' </summary>
  Private Sub New()
  End Sub

  #End Region

End Class

File: src/IsbInterop/IsbInteropConfiguration.cs
using System;
using System.ComponentModel;
using System.Configuration;
using System.IO;

namespace IsbInterop
{
  /// <summary>
  /// Конфигурация IsbInterop.
  /// </summary>
  public static class IsbInteropConfiguration
  {
    #region Поля и свойства

    /// <summary>
    /// Таймаут выполнения метода IS-Builder, секунд.
    /// </summary>
    public static int IsbMethodExecutionTimeout => _isbMethodExecutionTimeout.Value;

    private static readonly Lazy<int> _isbMethodExecutionTimeout =
      new Lazy<int>(() =>
      {
        const int defaultMethodExecutionTimeout = 60;
        return GetAppSettingOrDefault("IsbMethodExecutionTimeout", defaultMethodExecutionTimeout, t => t > 0);
      }, true);

    /// <summary>
    /// Таймаут создания приложения IS-Builder, секунд.
    /// </summary>
    public static int IsbAppCreationTimeout => _isbAppCreationTimeout.Value;

    private static readonly Lazy<int> _isbAppCreationTimeout =
      new Lazy<int>(() =>
      {
        const int defaultIsbAppCreationTimeout = 300;
        return GetAppSettingOrDefault("IsbAppCreationTimeout", defaultIsbAppCreationTimeout, t => t > 0);
      }, true);

    /// <summary>
    /// Конфигурация сборки IsbInterop.
    /// </summary>
    private static readonly Lazy<Configuration> StandaloneConfiguration =
      new Lazy<Configuration>(TryGetStandaloneConfiguration, true);

    #endregion

    #region Методы

    /// <summary>
    /// Шифрует строки подключения.
    /// </summary>
    public static void EncryptConnectionStrings()
    {
      var configuration = StandaloneConfiguration.Value;
      if (configuration == null)
      {
        try
        {
          configuration = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
        }
        catch (ConfigurationErrorsException)
        {
          return;
        }
      }

      var configSection = configuration.ConnectionStrings;

      if (configSection == null)
        return;

      try
      {
        if (!configSection.IsReadOnly() && !configSection.ElementInformation.IsLocked && !configSection.SectionInformation.IsLocked)
        {
          if (!configSection.SectionInformation.IsProtected)
            configSection.SectionInformation.ProtectSection("DataProtectionConfigurationProvider");

          configSection.SectionInformation.ForceSave = true;

          configuration.Save();
        }
      }
      // ReSharper disable once EmptyGeneralCatchClause
      catch (Exception)
      {
      }
    }

    /// <summary>
    /// Получает настройку секции appSettings.
    /// </summary>
    /// <typeparam name="T">Тип результата.</typeparam>
    /// <param name="settingName">Имя настройки/</param>
    /// <param name="result">Результат.</param>
    /// <returns>True, если удалось получить настройку, иначе false.</returns>
    public static bool TryGetAppSetting<T>(string settingName, ref T result)
    {
      if (settingName == null)
        throw new ArgumentNullException(nameof(settingName));

      string settingValue = null;

      var configuration = StandaloneConfiguration.Value;

      if (configuration != null)
      {
        var setting = configuration.AppSettings.Settings[settingName];
        if (setting != null)
          settingValue = setting.Value;
      }
      else
        settingValue = ConfigurationManager.AppSettings[settingName];

      if (settingValue != null)
        return TryConvert(settingValue, ref result);
      else
        return false;
    }

    /// <summary>
    /// Получает строку подключения.
    /// </summary>
    /// <returns>Настройки строки подключения.</returns>
    internal static string GetConnectionString()
    {
      ConnectionStringSettings connectionStringSettings;

      var configuration = StandaloneConfiguration.Value;

      if (configuration != null)
        connectionStringSettings = configuration.ConnectionStrings.ConnectionStrings["IsbInterop"];
      else
        connectionStringSettings = ConfigurationManager.ConnectionStrings["IsbInterop"];

      var connectionString = connectionStringSettings?.ConnectionString;

      return connectionString;
    }

    /// <summary>
    /// Получает конфигурацию сборки IsbInterop.
    /// </summary>
    /// <returns>Конфигурация, если она существует, иначе null.</returns>
    private static Configuration TryGetStandaloneConfiguration()
    {
      var currentAssembly = typeof(IsbApplicationManager).Assembly;
      var assemblyPath = new Uri(currentAssembly.CodeBase).LocalPath;
      var assemblyPathWithoutExtenstion = assemblyPath.Substring(0, assemblyPath.Length - 4);
      var configPath = $"{assemblyPathWithoutExtenstion}.dll.config";

      // Проверяем, что файл существует, т.к.метод OpenMappedExeConfiguration всегда создает объект конфигурации.
      if (!File.Exists(configPath))
        return null;

      var configMap = new ExeConfigurationFileMap();
      configMap.ExeConfigFilename = configPath;

      Configuration configuration = null;
      try
      {
        configuration = ConfigurationManager.OpenMappedExeConfiguration(configMap, ConfigurationUserLevel.None);
      }
      catch (ConfigurationErrorsException)
      {
        // We could log.trace this exception if we have logger.
      }

      return configuration;
    }

    /// <summary>
    /// Конвертирует строку в заданный тип.
    /// </summary>
    /// <typeparam name="T">Тип результата.</typeparam>
    /// <param name="source">Исходное значение.</param>
    /// <param name="result">Результат.</param>
    /// <returns>Экземпляр типа.</returns>
    private static bool TryConvert<T>(string source, ref T result)
    {
      var typeConverter = TypeDescriptor.GetConverter(typeof(T));
      if (typeConverter.CanConvertFrom(typeof(string)))
      {
        try
        {
          result = (T)typeConverter.ConvertFrom(source);
          return true;
        }
        // ReSharper disable once EmptyGeneralCatchClause
        catch (Exception)
        {
        }
      }

      return false;
    }

    /// <summary>
    /// Получает значение настройки из секции AppSettings.
    /// </summary>
    /// <typeparam name="T">Тип значения.</typeparam>
    /// <param name="setting">Настройка.</param>
    /// <param name="defaultValue">Значение по умолчанию.</param>
    /// <param name="predicate">Предикат.</param>
    /// <returns>Значение настройки.</returns>
    private static T GetAppSettingOrDefault<T>(string setting, T defaultValue, Func<T, bool> predicate = null)
    {
      T value = defaultValue;

      if (TryGetAppSetting(setting, ref value))
      {
        if (predicate != null && predicate(value))
          return value;
      }

      return defaultValue;
    }

    #endregion
  }
}
